src/TodosContext.tsx

```diff
@@ -0,0 +1,209 @@
+  setTempTodo: () => {},
+});
+
+interface ProviderProps {
+  children: React.ReactNode,
+}
+
+export const TodosProvider: React.FC<ProviderProps> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterBy, setFilterBy] = useState(Filter.ALL);
+  const [areCompletedDeletingNow, setAreCompletedDeletingNow]
+    = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const visibleTodos = filterTodos(todos, { filterBy });
+
+  useEffect(() => {
+    getTodos(USER_ID).then(setTodos);
+  }, []);
+
+  const handleErrorOccuring = (errorTitle: string) => {
+    setErrorMessage(errorTitle);
+
+    setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  };
+
+  const todoAdd = (newQuery: string) => {
+    const normalizedQuery = newQuery.trim();
+
+    return new Promise<boolean>((resolve, reject) => {
+      if (!normalizedQuery) {
+        handleErrorOccuring('Title can\'t be empty');
+        reject(new Error(errorMessage));
+
+        return;
+      }
+
+      const newTodo: Omit<Todo, 'id'> = {
+        userId: USER_ID,
+        title: normalizedQuery,
+        completed: false,
+      };
+
+      setTempTodo({ ...newTodo, id: 0 });
+
+      addTodo(newTodo)
+        .then((data) => {
+          setTodos([...todos, data]);
+          resolve(true);
+        })
+        .catch(() => {
+          handleErrorOccuring('Unable to add a todo');
+          reject(new Error(errorMessage));
+        });
+    })
+      .catch(error => error);
+  };
+
+  const todoDelete = (todoId: number) => {
+    return new Promise<boolean>((resolve, reject) => {
+      deleteTodo(todoId)
+        .then(() => {
+          setTodos(prevTodos => prevTodos.filter(todo => {
+            return todo.id !== todoId;
+          }));
+
+          resolve(true);
+        })
+        .catch(() => {
+          handleErrorOccuring('Unable to delete a todo');
+          reject(new Error(errorMessage));
+        });
+    })
+      .catch(error => error);
+  };
+
+  const clearAllCompleted = () => {
+    const todosIdsToOperate = [...todos]
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setAreCompletedDeletingNow(true);
+
+    deleteAllTodos(todosIdsToOperate)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => {
+          return !todosIdsToOperate.includes(todo.id);
+        }));
+        setAreCompletedDeletingNow(false);
+      });
+  };
+
+  const isTodosHasCompleted = () => {
+    return todos.some(todo => todo.completed);
+  };
+
+  const isEveryTodoCompleted = () => {
+    return todos.every(todo => todo.completed);
+  };
+
+  const value = {
+    todos,
+    visibleTodos,
+    todoAdd,
+    clearAllCompleted,
+    todoDelete,
+    isTodosHasCompleted,
+    isEveryTodoCompleted,
+    filterBy,
+    setFilterBy,
+    errorMessage,
+    areCompletedDeletingNow,
+    tempTodo,
+    setTempTodo,
+  };
```

Comment:
The 'value' object passed as the value prop to the Context provider changes every render. 
Might be better to consider wrapping it in a useMemo and moving setState functions to a different provider, they wouldn't change each render